### PR TITLE
Replay Settings

### PR DESCRIPTION
**Pull request by EVAST9919**
screenshot-120

### CODE REVIEW

File: osu.Game/Screens/Play/ReplaySettings/ReplaySettingsOverlay.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        private const int fade_duration = 100;
+
+        private bool isVisible;
+
+        private bool isAvaliable;
+        /// <summary>
+        /// Allow user to interact with this overlay.
+        /// </summary>
+        public bool IsAvaliable
```
Comment (smoogipooo) [inline osu.Game/Screens/Play/ReplaySettings/ReplaySettingsOverlay.cs]: How about doing `ReplaySettingsOverlay.Hide()` from inside `HUDOverlay.BindHitRenderer()`? Similar to how ModDisplay is done. Would remove this property + tidy up any usages.

Would also do the same for the key bind I think, further removing toggleVisibility, keeping this a pretty basic class.

File: osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+                    Children = new Drawable[]
+                    {
+                        new Container
+                        {
+                            Name = @"Header",
+                            RelativeSizeAxes = Axes.X,
+                            Height = header_height,
+                            Origin = Anchor.TopCentre,
+                            Anchor = Anchor.TopCentre,
```
Comment (smoogipooo) [inline osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs]: Prefer Origins + Anchors at the top of the initializer for consistency (there are other cases in this PR where this is not the case).

File: osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                            Padding = new MarginPadding(15),
+                            Spacing = new Vector2(0, 15),
+                        }
+                    }
+                },
+            };
+        }
+
+        public new void Add(Drawable drawable)
```
Comment (smoogipooo) [inline osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs]: `protected override Container<Drawable> Content => content`, change `Children =` to `InternalChildren =` and then this method can be removed.

File: osu.Game/Graphics/UserInterface/SimpleButton.cs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            Anchor = Anchor.Centre;
+
+            Children = new Drawable[]
+            {
+                content = new Container
+                {
+                    Size = new Vector2(button_size),
+                    CornerRadius = 5,
+                    Masking = true,
+
```
Comment (smoogipooo) [inline osu.Game/Graphics/UserInterface/SimpleButton.cs]: No space.

File: osu.Game/Screens/Play/ReplaySettings/ReplaySettingsOverlay.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            {
+                case Key.H:
+                    toogleVisibility();
+                    return true;
+            }
+
+            return base.OnKeyDown(state, args);
+        }
+
+        private void toogleVisibility()
```
Comment (smoogipooo) [inline osu.Game/Screens/Play/ReplaySettings/ReplaySettingsOverlay.cs]: `toggleVisibility`

File: osu.Game/Screens/Play/ReplaySettings/CollectionSettings.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+namespace osu.Game.Screens.Play.ReplaySettings
+{
+    public class CollectionSettings : SettingsDropdownContainer
+    {
+        protected override string Title => @"collections";
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            Add(new OsuSpriteText
```
Comment (smoogipooo) [inline osu.Game/Screens/Play/ReplaySettings/CollectionSettings.cs]: With my suggestion of `Content => content` in the other review comment, you can tidy this up using
`Children = new Drawable[] { ... };`

**Review by smoogipoo — CHANGES_REQUESTED** [review]
First code-only pass.

**Comment by smoogipooo** [comment]
I think we want to leave this in visual tests for now, until some of these options are implemented, then enable them selectively. We're trying not to have options in the game which aren't implemented yet.

File: osu.Game/Screens/Play/ReplaySettings/DiscussionSettings.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        protected override string Title => @"discussions";
+
+        [BackgroundDependencyLoader]
+        private void load(OsuConfigManager config)
+        {
+            Children = new Drawable[]
+            {
+                new ReplaySettingsCheckbox
+                {
+                    LabelText = "Show floating coments",
```
Comment (smoogipooo) [inline osu.Game/Screens/Play/ReplaySettings/DiscussionSettings.cs]: comments*

File: osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = Color4.Black,
+                    Alpha = 0.5f,
+                },
+                new FillFlowContainer
+                {
+                    Direction = FillDirection.Vertical,
+                    RelativeSizeAxes = Axes.X,
+                    AutoSizeAxes = Axes.Y,
+
```
Comment (smoogipooo) [inline osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs]: No whitespace in object initializer.

File: osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        private const float transition_duration = 600;
+        private const int container_width = 270;
+        private const int border_thickness = 2;
+        private const int header_height = 30;
+        private const int corner_radius = 5;
+
+        private FillFlowContainer content;
+        private SimpleButton button;
+        private bool buttonIsActive;
```
Comment (smoogipooo) [inline osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs]: buttonIsActive -> expanded

File: osu.Game/Graphics/UserInterface/SimpleButton.cs [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+using osu.Framework.Allocation;
+using osu.Framework.Extensions.Color4Extensions;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Input;
+
+namespace osu.Game.Graphics.UserInterface
+{
+    public class SimpleButton : ClickableContainer
```
Comment (smoogipooo) [inline osu.Game/Graphics/UserInterface/SimpleButton.cs]: Consider renaming this to IconButton.

File: osu.Game/Screens/Play/HUDOverlay.cs [hunk 11]
```diff
@@ -95,8 +99,13 @@ public virtual void BindHitRenderer(HitRenderer hitRenderer)
 
             // in the case a replay isn't loaded, we want some elements to only appear briefly.
             if (!hitRenderer.HasReplayLoaded)
+            {
+                ReplaySettingsOverlay.Hide();
+                ReplaySettingsOverlay.AlwaysPresent = false;
```
Comment (smoogipooo) [inline osu.Game/Screens/Play/HUDOverlay.cs]: What is the AlwaysPresent value here for?

File: osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+using OpenTK;
+using OpenTK.Graphics;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Graphics.Sprites;
+
+namespace osu.Game.Graphics.UserInterface
```
Comment (smoogipooo) [inline osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs]: I think it's best if this resided inside Screens/Play/ReplaySettings for now.

File: osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+{
+    public abstract class SettingsDropdownContainer : Container
```
Comment (smoogipooo) [inline osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs]: ReplaySettingsGroup.

File: osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    {
+        /// <summary>
+        /// The title of this container, which will be written in header.
```
Comment (smoogipooo) [inline osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs]: If class is renamed to ReplaySettingsGroup -> "The title to be displayed in the header of this group."

File: osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+        private FillFlowContainer content;
+        private SimpleButton button;
+        private bool buttonIsActive;
+
+        private Color4 buttonActiveColour;
+
+        [BackgroundDependencyLoader]
+        private void load(OsuColour colours)
+        {
```
Comment (smoogipooo) [inline osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs]: All of this method should be moved to the constructor, after which `content` and `button` can be made readonly. Use load() for initializing the button colour.

Generally I think we should keep ctors for object initialization and load() for virtual methods + DI initialization. Having this stuff inside load() messes with object initializers because the load() method would override them - possibly unexpectedly.

File: osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+                    Children = new Drawable[]
+                    {
+                        new Container
+                        {
+                            Name = @"Header",
+                            Origin = Anchor.TopCentre,
+                            Anchor = Anchor.TopCentre,
+                            RelativeSizeAxes = Axes.X,
+                            Height = header_height,
+
```
Comment (smoogipooo) [inline osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs]: And here.

File: osu.Game/Screens/Play/ReplaySettings/CollectionSettings.cs [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            Children = new Drawable[]
+            {
+                new OsuSpriteText
+                {
+                    Text = @"Add current song to",
+                },
+                new CollectionsDropdown<PlaylistCollection>
```
Comment (smoogipooo) [inline osu.Game/Screens/Play/ReplaySettings/CollectionSettings.cs]: We'll need `CollectionsDropdown : SettingsDropdown` to provide the label for us, otherwise the label and the dropdown are too far apart from one another. Check how I handled  this with ReplaySettingsCheckbox/ReplaySettingsSliderBar.

File: osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs (hunk 15 above)
Comment (smoogipooo) [inline osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs]: And make sure to move every other class from load() to ctor along with this.

**Review by smoogipoo — DISMISSED** [review]
See comments.

File: osu.Game/Screens/Play/HUDOverlay.cs (hunk 11 above)
Comment (EVAST9919) [inline osu.Game/Screens/Play/HUDOverlay.cs]: We need to block toggling visibility if we are not in replay

Comment (smoogipooo) [inline osu.Game/Screens/Play/HUDOverlay.cs]: That's why I suggested putting the key event for toggling visibility in this class' OnKeyDown.

File: osu.Game/Overlays/Settings/SettingsDropdown.cs [hunk 18]
```diff
@@ -27,7 +27,7 @@ public class SettingsDropdown<T> : SettingsItem<T>
             }
         }
 
-        protected override Drawable CreateControl() => dropdown = new OsuDropdown<T>
+        protected override Drawable CreateControl() => dropdown = new Graphics.UserInterface.SettingsDropdown<T>
```
Comment (smoogipooo) [inline osu.Game/Overlays/Settings/SettingsDropdown.cs]: Is this change intended?

File: osu.Game/Graphics/UserInterface/OsuDropdown.cs [hunk 19]
```diff
@@ -13,7 +13,7 @@
 
 namespace osu.Game.Graphics.UserInterface
 {
-    public class OsuDropdown<T> : Dropdown<T>
+    public class SettingsDropdown<T> : Dropdown<T>
```
Comment (smoogipooo) [inline osu.Game/Graphics/UserInterface/OsuDropdown.cs]: Guess that goes along with this, I don't think this change is intended :P

**Comment by peppy** [comment]
https://streamable.com/ul5dk

This feels really wrong

File: osu.Game/Screens/Play/ReplaySettings/ReplaySettingsGroup.cs [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+
+        private const float transition_duration = 600;
+        private const int container_width = 270;
+        private const int border_thickness = 2;
+        private const int header_height = 30;
+        private const int corner_radius = 5;
+
+        private readonly FillFlowContainer content;
+        private readonly IconButton button;
+        private bool expanded;
```
Comment (peppy) [inline osu.Game/Screens/Play/ReplaySettings/ReplaySettingsGroup.cs]: this is always opposite of what is should be...

**Comment by peppy** [comment]
I'm going to merge this with my changes. There needs to be changes to the underlying `FillFlow` implementation to fix the animation issues.

I've stopped this from being displayed outside of the VisualTest as it adds no functionality to the game. It can be added once it has functionality.

**Review by peppy — DISMISSED** [review]

**Review by peppy — APPROVED** [review]
